### PR TITLE
LR Finder: compute gradients w.r.t. log10(lr) in exponential mode (fix torch.gradient spacing)

### LINKED ISSUE #21141: [BUG] Learning Rate finder does not find optimal lr anymore since version 2.5.3
### Bug description

Hi all and thanks for the awesome library 🚀 
 
Since #21055 was merged (part of the version 2.5.3 release), the Learning Rate finders do not find the optimal learning rate anymore with noisy data. 

Steps to reproduce (you can run below once with python-lightning<2.5.3 and once with >=2.5.3) are below. The example shows the lr finder behavior on synthetic noise-free and noisy losses.

```py
import matplotlib.pyplot as plt
import numpy as np

from pytorch_lightning.callbacks.lr_finder import _LRFinder

use_lightning_le_2_5_3 = True

np.random.seed(42)
n_lrs = 100

# generate losses (noise-free and noisy)
x = np.linspace(-50, 50, n_lrs)
losses = (-x**2 - 30*x + 4000) / 1000.
noise = np.random.uniform(-0.1, 0.1, n_lrs)
losses_noisy = losses + noise

# generate exponential lrs (as done in _LRFinder)
lr_min = 1e-8
lr_max = 1.
lrs = []
for i in range(n_lrs):
    r = (i + 1) / n_lrs
    lrs.append(lr_min * (lr_max / lr_min) ** r)

fig, axes = plt.subplots(ncols=2, figsize=(10, 4))
for ax, losses_, name in zip(
    axes,
    [losses, losses_noisy],
    ["noise-free", "noisy"],
):
    # fill in the results manually to simulate a lr_finder run without actual training
    lr_finder = _LRFinder(mode="exponential", lr_min=lr_min, lr_max=lr_max, num_training=n_lrs)
    lr_finder.results = {
        "lr": lrs,
        "loss": losses_
    }
    # plot with the suggested lr
    lr_finder.plot(suggest=True, ax=ax)
    ax.set_title(f"LR Finder ({name})")

plt.suptitle(f"pytorch-lightning{'<2.5.3' if use_lightning_le_2_5_3 else '>=2.5.3'}")
fig.tight_layout()
plt.show(block=True)
```

This will give you the following suggested lrs:

Image

Image

As you can see, before 2.5.3, the results weren't affected by noise in the data. After 2.5.3 it does result in suggesting a non-optimal lr in the lower learning rates, since the gradients using the newly introduced spacing have much larger magnitude in these areas. 

Having noise in the losses is something that comes naturally in our actual use cases. I'm from the [Darts forecasting library](https://github.com/unit8co/darts) where we built our neural network framework around lightning.

### Expected behavior

The LRFinder suggestions should ideally not be affected by minor noise in the data (of course with large noise that's a different story).

### Proposed solutions

That's a good question.. Maybe we could de-noise / filtering / moving average the loss data in the LRFinder  before the gradients are computed?

Thanks again, and if you need more info, let me know!

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
* CUDA:
	- GPU:               None
	- available:         False
	- version:           None
* Lightning:
	- lightning:         2.5.3
	- lightning-utilities: 0.15.2
	- pytorch-lightning: 2.5.2
	- torch:             2.8.0
	- torchmetrics:      1.8.1
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
		- 
	- processor:         arm
	- python:            3.12.11
	- release:           24.3.0
	- version:           Darwin Kernel Version 24.3.0: Thu Jan  2 20:24:16 PST 2025; root:xnu-
* installed with `pip`
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by KAVYANSHTYAGI**
Fixes #21141

**Summary**

- Fix LR-Finder’s suggestion by giving torch.gradient the correct sample points:
- Exponential mode: differentiate w.r.t. log10(lr) → spacing=[torch.log10(lrs)]
- Linear mode: differentiate w.r.t. lr → spacing=[lrs]
- This matches how points are sampled and avoids bias from treating log-spaced LRs as linear.

**Motivation**

> torch.gradient(y, spacing=x)

expects the coordinates of samples. With log-spaced LRs, the intended derivative is 
∂loss/∂log⁡10lr
Previous behavior used linear LR coordinates, which is inconsistent and can skew the suggested LR.

**Tests**

- Use torch.logspace for exponential mode.
- Assert finite gradients and correct index selection using proper spacing.
- Remove brittle “with vs without spacing must differ” check.

**Impact**

- User-visible: Suggested LR may shift in mode="exponential" (now mathematically correct).
- API: No changes.
- Performance: Negligible overhead (single log10 over ~10²–10³ points).

----
📚 Documentation preview 📚: https://pytorch-lightning--21171.org.readthedocs.build/en/21171/

### CODE REVIEW

**Comment by SkafteNicki** [comment]
@KAVYANSHTYAGI thanks for the PR. Does this solve the problems mention in issue https://github.com/Lightning-AI/pytorch-lightning/issues/21141?

**Review by KAVYANSHTYAGI — COMMENTED** [review]
It mitigates the problem from #21141 but does not fully implement the “step based gradient” behavior some commenters requested.

**What this PR fixes**

- In mode="exponential", gradients are now computed w.r.t. log10(lr) (i.e., the actual sampling coordinate) not linear LR.
- This removes the non-uniform-grid pathology introduced in ≥2.5.3, which amplified early-step noise and biased suggestions toward tiny LRs.

**What it does not change**

- It does not switch the derivative to training-step index (uniform step spacing). Linear mode remains unchanged.

**Review by Borda — APPROVED** [review]

**Comment by SkafteNicki** [comment]
Got confirmation that issue #21141 can be closed when this is merged, so we can move forward with this

**Review by SkafteNicki — APPROVED** [review]

**Comment by daidahao** [comment]
With all due respect, this essentially reverts #21055 and is mathematically equivalent to how gradient was calculated before. This is also what I suggested in #21141. To be clear, I am not blaming anyone here. Rather, I think we could all learn from this experience that robustly finding a "good" learning rate is a hard problem deserving in-depth discussions and real-world testing before code breaking changes. In light of this, I would recommend that any future changes of LR finder behaviours should require a good amount of academic literature and/or empirical evidence.

@SkafteNicki @dennisbader @KAVYANSHTYAGI Thank you for your contributions on addressing this issue so quickly.

**Comment by dennisbader** [comment]
Hi @SkafteNicki, I just wanted to check in about the status of this PR. Is there anything I can help to get it merged?

**Comment by SkafteNicki** [comment]
> Hi @SkafteNicki, I just wanted to check in about the status of this PR. Is there anything I can help to get it merged?

@dennisbader sorry for the delay, our CI is very flaky at the moment so everything is kind of on hold

**Comment by dennisbader** [comment]
Hi @SkafteNicki, no worries :) It seems like many other PRs have successfully passed / been merged in the meantime. Do you think we could give it another try?

**Comment by daidahao** [comment]
@SkafteNicki Can you give an update on this please? Without the fix here or a revert of #21055, Darts has to cap lightning version to be `<2.5.3`.

**Comment by dennisbader** [comment]
Sorry for bothering you guys, and well aware that this might not top-priority @Borda @SkafteNicki. Since this has been stale for almost a year now, I was hoping to get an update on this PR. Do you think this will ever get merged, or should we start finding workarounds?

If there are any leftover topics, or I can assist in any way to get this through, let me know :)

**Comment by Borda** [comment]
> Sorry for bothering you guys, and well aware that this might not top-priority @Borda @SkafteNicki. 

I guess it is but you need to ping maintainers...
